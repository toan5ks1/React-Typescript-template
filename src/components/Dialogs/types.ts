import { ReactElement, ReactNode } from "react";

import { UseFormMethods } from "src/typings/react-hook-form";

import { Breakpoint } from "@mui/material";
import { ButtonBaseProps } from "src/components/Buttons/ButtonBase";

import { WrapperPageContentProps } from "../Wrappers/WrapperPageContent";
import { DialogBaseProps } from "./DialogBase";

export type DialogFooterType = ReactElement | "portal" | "empty" | "default";
export interface DialogWithHeaderFooterProps extends DialogBaseProps {
    onEdit?: () => void;
    onDelete?: () => void;
    onClose: () => void;
    onSave?: () => void;
    onCloseClick?: () => void;
    maxWidthBox?: Breakpoint;
    minWidthBox?: Breakpoint;
    header?: ReactElement;
    footer?: DialogFooterType;
    textClose?: string;
    textSave?: string;
    footerConfirmButtonProps?: {
        variant?: ButtonBaseProps["variant"];
        color?: ButtonBaseProps["color"];
        disabled?: ButtonBaseProps["disabled"];
    };
    shouldDisableClosing?: boolean;
    shouldShowCancelButton?: boolean;
    additionalHeights?: number;
    shouldPressKey?: boolean;
}

export interface DialogFullScreenProps extends DialogBaseProps {
    onSave?: () => void;
    onClose: () => void;
    textClose?: string;
    textSave?: string;
    footerConfirmButtonProps?: {
        variant?: ButtonBaseProps["variant"];
        color?: ButtonBaseProps["color"];
        disabled?: ButtonBaseProps["disabled"];
    };
    footer?: ReactElement;
    title: string;
    isShowingBackdrop?: boolean;
    contentSize?: WrapperPageContentProps["variant"];
    dialogCancelConfirmProps?:
        | Pick<DialogCancelConfirmProps, "title" | "textCancelDialog" | "textSave">
        | boolean;
}

export type DialogFooterProps = Pick<
    DialogWithHeaderFooterProps | DialogFullScreenProps,
    "onClose" | "onSave" | "textClose" | "textSave" | "footerConfirmButtonProps"
>;

export interface DialogHeaderProps
    extends Pick<
        DialogWithHeaderFooterProps,
        "onClose" | "onDelete" | "onEdit" | "title" | "onCloseClick" | "shouldDisableClosing"
    > {
    children?: ReactNode;
    disableBottomPadding?: boolean;
}

export interface DialogCancelConfirmProps
    extends Pick<
        DialogWithHeaderFooterProps,
        | "onSave"
        | "title"
        | "onClose"
        | "open"
        | "textClose"
        | "textSave"
        | "footerConfirmButtonProps"
    > {
    textCancelDialog?: ReactNode;
}

export type DialogFooterConfirmProps = Pick<
    DialogWithHeaderFooterProps,
    | "onClose"
    | "onSave"
    | "textClose"
    | "textSave"
    | "footerConfirmButtonProps"
    | "shouldShowCancelButton"
>;

export type DialogFullScreenHeaderProps = Pick<DialogFullScreenProps, "onClose" | "title">;

export interface FooterButtonProps {
    variant?: ButtonBaseProps["variant"];
    color?: ButtonBaseProps["color"];
}

export interface DialogHFProps<T> extends DialogWithHeaderFooterProps, UseFormMethods<T> {}

export interface DialogFullScreenHFProps<T> extends DialogFullScreenProps, UseFormMethods<T> {}
