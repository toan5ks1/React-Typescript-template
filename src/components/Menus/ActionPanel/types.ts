import { ReactElement, ReactNode } from "react";

import { MutationMenus } from "src/common/constants/enum";
import { TypeEntity } from "src/typings/react-admin";

import { MenuProps } from "@mui/material";
import { IconButtonWithAnchorMenuProps } from "src/components/IconButton/IconButtonWithAnchorMenu";

export interface CustomAction<T = MutationMenus> {
    action: T;
    Icon?: ReactElement;
    label?: string;
    withConfirm?: boolean;
    className?: string;
    disabled?: boolean;
    labelOptions?: object;
    onClick?: () => void;
    render?: (label: string, onAction: () => void) => ReactNode;
}

export type Action<T = MutationMenus> = MutationMenus | CustomAction<T>;

export interface GetActionPropsReturn<T = MutationMenus>
    extends Omit<CustomAction<T>, "action" | "Icon"> {
    actionType: T;
}

export interface ActionPanelProps<T extends any = any, TAction = MutationMenus>
    extends Omit<MenuProps, "onClick" | "open"> {
    size?: "inherit" | "default" | "small" | "large";
    record?: T;
    loading?: boolean;
    disables?: {
        [x: string]: boolean;
    };
    disabled?: boolean;
    onAction?: (...args: any[]) => void;
    recordName: ReactNode;
    actions?: Action<TAction>[];
    resource?: TypeEntity | string;
    buttonStyle?: IconButtonWithAnchorMenuProps["variant"];
    suffixDeleteTitle?: string;
    getActionProps?: (action: Action<TAction>) => GetActionPropsReturn<TAction>;
}

export interface WrapperActionPanelProps<T extends any = any, TAction = MutationMenus>
    extends Omit<ActionPanelProps<T, TAction>, "onClick"> {
    open: boolean;
    anchorEl: HTMLElement | null;
    onClose: () => void;
    onAction: (e: string) => void;
    menuProps?: MenuProps;
    onClick?: () => void;
}
