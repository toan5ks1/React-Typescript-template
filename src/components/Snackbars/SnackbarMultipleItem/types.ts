export type CloseReasonTypes = "timeout" | "maxsnack" | "instructed";
export type SnackbarKey = string;
export type SnackbarMessage = string;

export type TransitionCloseHandler = (
    event: React.SyntheticEvent<any> | null,
    reason: CloseReasonTypes,
    key?: SnackbarKey
) => void;

export type TransitionHandler = (node: HTMLElement, key: SnackbarKey) => void;

export interface TransitionHandlerProps {
    onClose: TransitionCloseHandler;
    onExited: TransitionHandler;
}
