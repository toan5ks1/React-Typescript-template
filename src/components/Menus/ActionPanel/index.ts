export { default } from "./ActionPanel";
export * from "./types";
export * from "./helpers";
