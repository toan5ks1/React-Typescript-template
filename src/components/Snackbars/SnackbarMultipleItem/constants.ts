export enum CloseReason {
    TIMEOUT = "timeout",
    MAXSNACK = "maxsnack",
    INSTRUCTED = "instructed",
}

export enum DefaultsSnackbar {
    MAX_SNACK = 3,
    VARIANT = "success",
    AUTO_HIDE_DURATION = 5000,
}
