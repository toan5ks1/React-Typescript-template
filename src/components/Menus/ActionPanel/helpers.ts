import { merge } from "lodash";
import { MutationMenus } from "src/common/constants/enum";

import { Action, GetActionPropsReturn } from "./types";

//Action options
export const AllowedActions: { [x: string]: GetActionPropsReturn<MutationMenus> } = {
    [MutationMenus.DUPLICATE]: {
        actionType: MutationMenus.DUPLICATE,
        label: "ra.common.action.duplicate",
    },
    [MutationMenus.MOVE_UP]: {
        actionType: MutationMenus.MOVE_UP,
        label: "ra.common.action.moveUp",
    },
    [MutationMenus.MOVE_DOWN]: {
        actionType: MutationMenus.MOVE_DOWN,
        label: "ra.common.action.moveDown",
    },
    [MutationMenus.EDIT]: {
        actionType: MutationMenus.EDIT, //should be same as the object key
        label: "ra.common.action.edit",
    },
    [MutationMenus.DELETE]: {
        actionType: MutationMenus.DELETE,
        label: "ra.common.action.delete",
        withConfirm: true,
    },
    [MutationMenus.DETAIL]: {
        actionType: MutationMenus.DETAIL,
        label: "ra.common.action.detail",
    },
    [MutationMenus.DOWNLOAD]: {
        actionType: MutationMenus.DOWNLOAD,
        label: "ra.common.action.download",
        labelOptions: { smart_count: 2, extension: "CSV" },
    },
    [MutationMenus.UPLOAD_FILE]: {
        actionType: MutationMenus.UPLOAD_FILE,
        label: "ra.common.action.uploadFile",
    },
    [MutationMenus.RENAME]: {
        actionType: MutationMenus.RENAME,
        label: "ra.common.action.rename",
    },
    [MutationMenus.UPDATE]: {
        actionType: MutationMenus.UPDATE,
        label: "ra.common.action.update",
    },
    [MutationMenus.ARCHIVE]: {
        actionType: MutationMenus.ARCHIVE,
        label: "ra.common.action.archive",
    },
    [MutationMenus.UNARCHIVE]: {
        actionType: MutationMenus.UNARCHIVE,
        label: "ra.common.action.unarchive",
    },
    [MutationMenus.RE_ISSUE_PASSWORD]: {
        actionType: MutationMenus.RE_ISSUE_PASSWORD,
        label: "ra.common.action.reIssuePassword",
    },
    [MutationMenus.REMOVE]: {
        actionType: MutationMenus.REMOVE,
        label: "ra.common.action.remove",
    },
    [MutationMenus.PRINT_AS_STUDENT_CARD]: {
        actionType: MutationMenus.PRINT_AS_STUDENT_CARD,
        label: "ra.common.action.printAsStudentCard",
    },
    [MutationMenus.PRINT_AS_QR_CODE_SHEET]: {
        actionType: MutationMenus.PRINT_AS_QR_CODE_SHEET,
        label: "ra.common.action.printAsQrCodeSheet",
    },
    [MutationMenus.DOWNLOAD_STUDENT_QR_URLS]: {
        actionType: MutationMenus.DOWNLOAD_STUDENT_QR_URLS,
        label: "ra.common.action.downloadStudentQrUrls",
    },
    [MutationMenus.IMPORT_PARENTS]: {
        actionType: MutationMenus.IMPORT_PARENTS,
        label: "ra.common.action.importParents",
    },
    [MutationMenus.PAYMENT_CREATE_NEW_ORDER]: {
        actionType: MutationMenus.PAYMENT_CREATE_NEW_ORDER,
        label: "ra.common.action.createNewOrder",
    },
    [MutationMenus.PAYMENT_CREATE_ENROLLMENT_ORDER]: {
        actionType: MutationMenus.PAYMENT_CREATE_ENROLLMENT_ORDER,
        label: "ra.common.action.createEnrollmentOrder",
    },
    [MutationMenus.PAYMENT_CREATE_BULK_ORDER]: {
        actionType: MutationMenus.IMPORT_PARENTS,
        label: "ra.common.action.createBulkOrder",
    },
    [MutationMenus.PAYMENT_CREATE_CUSTOM_BILLING_ORDER]: {
        actionType: MutationMenus.PAYMENT_CREATE_CUSTOM_BILLING_ORDER,
        label: "ra.common.action.createCustomBillingOrder",
    },
};

export function getActionProps<T = MutationMenus>(action: Action<T>): GetActionPropsReturn<T> {
    const actionType = typeof action === "string" ? action : action.action;
    const fallback: GetActionPropsReturn<T> = {
        actionType: actionType as T,
        ...(typeof action === "string" ? {} : action),
    };
    const predefined = AllowedActions[actionType as string];

    if (predefined) {
        return merge({}, predefined, fallback); //If we don't use shadow object, it will be changed all allowedActions object
    }

    return merge(fallback, action);
}
