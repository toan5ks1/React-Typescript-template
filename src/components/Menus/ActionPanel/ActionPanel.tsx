import { cloneElement, useCallback, useEffect, useMemo, useRef, useState } from "react";

import { MutationMenus } from "src/common/constants/enum";

import { Box, Button, MenuItem, Popover } from "@mui/material";
import DialogWithHeaderFooter from "src/components/Dialogs/DialogWithHeaderFooter";
import IconButtonWithAnchorMenu from "src/components/IconButton/IconButtonWithAnchorMenu";
import TypographyTextSecondary from "src/components/Typographys/TypographyTextSecondary";

import { getActionProps as getActionPropsDefault } from "./helpers";
import { Action, ActionPanelProps, WrapperActionPanelProps } from "./types";

import isFunction from "lodash/isFunction";
import useActionMenu from "src/hooks/useActionMenu";
import useDialog from "src/hooks/useDialog";
import useTranslate from "src/hooks/useTranslate";

const sx = {
    menuItem: {
        minWidth: "100px",
        width: "100%",
    },
    text: {
        textTransform: "capitalize",
    },
};

export const WrapperActionPanel = (props: WrapperActionPanelProps) => {
    const {
        actions = [],
        loading,
        disables,
        onAction,
        onClose,
        open,
        anchorEl,
        menuProps = {},
        recordName,
        resource,
        suffixDeleteTitle,
        getActionProps,
    } = props;

    //TODO: refactor useTranslate for flexible to use useTranslate for all squads
    const t = useTranslate();
    const onClick = useCallback(
        (action: MutationMenus) => {
            onAction(action);
            onClose();
        },
        [onAction, onClose]
    );
    const { open: dialogOpen, onOpen: onDialogOpen, onClose: onDialogClose } = useDialog(false);
    const [currentAction, setCurrentAction] = useState<MutationMenus>();

    const suffixDeleteTitleDialog = useMemo(() => {
        if (suffixDeleteTitle) return suffixDeleteTitle;
        return resource ? t(`resources.${resource}.name`) : 1;
    }, [resource, suffixDeleteTitle, t]);

    return (
        <>
            <Popover
                anchorEl={anchorEl}
                anchorOrigin={{
                    vertical: "bottom",
                    horizontal: "right",
                }}
                transformOrigin={{
                    vertical: "top",
                    horizontal: "right",
                }}
                style={{ marginTop: "8px" }}
                keepMounted
                open={open}
                onClose={onClose}
                data-testid={open ? "ActionPanel__popover--open" : "ActionPanel__popover"}
                {...menuProps}
            >
                <Box p="8px 0" data-testid="ActionPanel__menuList">
                    {actions.map((action: Action) => {
                        const actionProps =
                            typeof getActionProps === "function"
                                ? getActionProps(action)
                                : getActionPropsDefault(action);
                        const {
                            actionType,
                            labelOptions,
                            onClick: onActionClick,
                            render,
                            label = "",
                            withConfirm,
                        } = actionProps;

                        const translatedLabel = t(label, labelOptions);
                        const iLoading = loading || (disables ? disables[actionType] : false);

                        let handleMenuItemClick = () => onClick(actionType);
                        let menuItemProps = {
                            "data-value": actionType,
                            component: Button,
                            sx: sx.menuItem,
                            variant: "text",
                            "aria-label": translatedLabel,
                            disabled: iLoading,
                            onClick: handleMenuItemClick,
                        };
                        let menuItem = (
                            <MenuItem key={actionType} {...menuItemProps}>
                                <Box component="span" sx={sx.text}>
                                    {translatedLabel}
                                </Box>
                            </MenuItem>
                        );

                        if (render) {
                            handleMenuItemClick = () => {
                                onActionClick && onActionClick();
                            };

                            return cloneElement(
                                menuItem,
                                { onClick: handleMenuItemClick },
                                render(translatedLabel, () => onClick(actionType))
                            );
                        }

                        if (withConfirm) {
                            return cloneElement(menuItem, {
                                onClick: () => {
                                    setCurrentAction(actionType);
                                    onDialogOpen();
                                },
                            });
                        }

                        return menuItem;
                    })}
                </Box>
            </Popover>
            <DialogWithHeaderFooter
                open={dialogOpen}
                title={t("ra.common.deleteDialogTitle", {
                    smart_count: suffixDeleteTitleDialog,
                })}
                textSave={t("ra.common.action.confirm")}
                onClose={onDialogClose}
                onSave={() => onClick(currentAction!)}
                footerConfirmButtonProps={{
                    color: "error",
                }}
            >
                <TypographyTextSecondary variant="body2">
                    {t("ra.common.deleteConfirmText", { smart_count: 1 })}&nbsp;
                    <strong>{recordName}</strong>&nbsp;
                    {t("ra.common.deleteConfirmText", { smart_count: 2 })}
                </TypographyTextSecondary>
            </DialogWithHeaderFooter>
        </>
    );
};

const ActionPanel = (props: ActionPanelProps) => {
    const { className, style, record, onAction, buttonStyle, loading, disabled, getActionProps } =
        props;
    const { open, onOpen, onClose } = useActionMenu();

    const anchorRef = useRef<HTMLDivElement>(null);
    const prevOpen = useRef(open);

    useEffect(() => {
        if (prevOpen.current && !open && anchorRef.current) {
            anchorRef.current.focus();
        }

        prevOpen.current = open;
    }, [open]);

    const _onAction = useCallback(
        (actionName) => {
            return isFunction(onAction) ? onAction(actionName, record) : undefined;
        },
        [onAction, record]
    );

    return (
        <Box display="flex" className={className} style={style} data-testid="ActionPanel__root">
            <IconButtonWithAnchorMenu
                ref={anchorRef}
                aria-haspopup="menu"
                sx={{ width: "40px", height: "40px" }}
                onClick={onOpen}
                data-testid="ActionPanel__trigger"
                variant={buttonStyle}
                loading={loading}
                disabled={disabled}
            />
            {open && (
                <WrapperActionPanel
                    {...props}
                    open={open}
                    anchorEl={anchorRef.current}
                    onClose={onClose}
                    onAction={_onAction}
                    getActionProps={getActionProps}
                />
            )}
        </Box>
    );
};

ActionPanel.defaultProps = {
    actions: [],
    loading: false,
    disables: {},
};

export default ActionPanel;
