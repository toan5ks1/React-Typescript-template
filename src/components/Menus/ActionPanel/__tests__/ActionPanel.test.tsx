import { MutationMenus } from "src/common/constants/enum";

import ActionPanel from "../ActionPanel";
import { ActionPanelProps, CustomAction } from "../types";

import { fireEvent, render, RenderResult, screen } from "@testing-library/react";
import TestApp from "src/test-utils/TestApp";
import TestThemeProvider from "src/test-utils/TestThemeProvider";

const renderUtil = (override: Partial<ActionPanelProps> = {}) => {
    const defaultProps: ActionPanelProps = {
        loading: false,
        actions: [],
        size: "inherit",
        recordName: <>Name</>,
        onClose: () => jest.fn(),
        onAction: () => jest.fn(),
    };

    const finalProps: ActionPanelProps = { ...defaultProps, ...override };

    return render(
        <TestApp>
            <TestThemeProvider>
                <ActionPanel {...finalProps} />
            </TestThemeProvider>
        </TestApp>
    );
};

describe("<ActionPanel /> close", () => {
    let wrapper: RenderResult;
    const customAction: CustomAction = {
        action: MutationMenus.DELETE,
        onClick: jest.fn(),
        render: (label) => <span>{label}</span>,
    };
    beforeEach(() => {
        wrapper = renderUtil({ actions: [MutationMenus.EDIT, customAction] });
    });

    it("should match snapshot", () => {
        expect(wrapper.container).toMatchSnapshot();
    });

    it("should render correct UI", () => {
        expect(screen.queryByTestId("ActionPanel__root")).toBeInTheDocument();
        expect(screen.queryByTestId("ActionPanel__trigger")).toBeInTheDocument();
    });

    it("should render custom action correctly", () => {
        fireEvent.click(screen.getByTestId("ActionPanel__trigger"));

        const deleteButton = screen
            .getByTestId("ActionPanel__popover--open")
            .querySelector("button:last-child")!;

        expect(deleteButton).toHaveTextContent("Delete");
        fireEvent.click(deleteButton);
        expect(customAction.onClick).toHaveBeenCalled();
    });
});

describe("Action panel for delete action", () => {
    it("should render suffix delete title when passed", () => {
        const suffixDeleteTitle = "Your name delete title";
        renderUtil({
            actions: [MutationMenus.DELETE],
            suffixDeleteTitle,
        });
        fireEvent.click(screen.getByTestId("ActionPanel__trigger"));

        const deleteButton = screen
            .getByTestId("ActionPanel__popover--open")
            .querySelector("button:last-child")!;

        fireEvent.click(deleteButton);
        screen.getByText(`Delete ${suffixDeleteTitle}`);
    });
});
