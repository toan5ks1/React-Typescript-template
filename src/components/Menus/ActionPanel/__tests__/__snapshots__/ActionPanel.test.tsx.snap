// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ActionPanel /> close should match snapshot 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-1 .IconButtonWithAnchorMenu-sizeSmall {
  padding: 10px;
}

.emotion-1.IconButtonWithAnchorMenu-root:hover {
  color: #2196F3;
}

.emotion-1 .IconButtonWithAnchorMenu-iconButton {
  border-radius: 8px;
  background-color: #EEEEEE;
}

.emotion-2 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  position: relative;
  box-sizing: border-box;
  -webkit-tap-highlight-color: transparent;
  background-color: transparent;
  outline: 0;
  border: 0;
  margin: 0;
  border-radius: 0;
  padding: 0;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  vertical-align: middle;
  -moz-appearance: none;
  -webkit-appearance: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: inherit;
  text-align: center;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  font-size: 1.5rem;
  padding: 8px;
  border-radius: 50%;
  overflow: visible;
  color: rgba(0, 0, 0, 0.54);
  -webkit-transition: background-color 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: background-color 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  padding: 12px;
  font-size: 1.75rem;
  width: 40px;
  height: 40px;
}

.emotion-2::-moz-focus-inner {
  border-style: none;
}

.emotion-2.Mui-disabled {
  pointer-events: none;
  cursor: default;
}

@media print {
  .emotion-2 {
    -webkit-print-color-adjust: exact;
    color-adjust: exact;
  }
}

.emotion-2:hover {
  background-color: rgba(0, 0, 0, 0.04);
}

@media (hover: none) {
  .emotion-2:hover {
    background-color: transparent;
  }
}

.emotion-2.Mui-disabled {
  background-color: transparent;
  color: #BDBDBD;
}

.emotion-3 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  width: 1em;
  height: 1em;
  display: inline-block;
  fill: currentColor;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  font-size: 1.25rem;
}

.emotion-4 {
  overflow: hidden;
  pointer-events: none;
  position: absolute;
  z-index: 0;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  border-radius: inherit;
}

<div>
  <div
    class="MuiBox-root emotion-0"
    data-testid="ActionPanel__root"
  >
    <div
      class="IconButtonWithAnchorMenu-root emotion-1"
    >
      <button
        aria-haspopup="menu"
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeLarge IconButtonWithAnchorMenu-root emotion-2"
        data-testid="ActionPanel__trigger"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall emotion-3"
          data-testid="MoreHorizIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M6 10c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm12 0c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm-6 0c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root emotion-4"
        />
      </button>
    </div>
  </div>
</div>
`;
